File: 141.-Linked-List-Cycle/step3.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
# 6m
class Solution:
    def hasCycle(self, head: ListNode | None) -> bool:
        fast: ListNode = head;
        slow: ListNode = head;
        while fast is not None:
            if fast.next is not None:
                fast = fast.next.next
                slow = slow.next
                if fast is slow:
                    return True
            else:
                return False
        return False

File: 141.-Linked-List-Cycle/step3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr){}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* fast = head;
        ListNode* slow = head;
        while(fast != nullptr){
            if(fast->next != nullptr){
                fast = fast->next->next;
                slow = slow->next;
                if(fast == slow){
                    return true;
                }
            }
            else{
                return false;
            }
        }
        return false;
    }
};